pkg/metrics/metrics.jsx

```diff
@@ -440,17 +474,68 @@ class CurrentMetrics extends React.Component {
         }
 
         // top 5 CPU and memory consuming systemd units
-        newState.topServicesCPU = n_biggest(this.cgroupCPUNames, this.samples[9], 5).map(
-            ([key, value, is_user, is_container]) => cgroupRow(key, Number(value / 10 / numCpu).toFixed(1), is_user, is_container) // usec/s → percent
+        const topServicesCPU = n_biggest(this.cgroupCPUNames, this.samples[9], 5);
+        newState.topServicesCPU = topServicesCPU.map(
+            ([key, value, is_user, is_container, userid]) => cgroupRow(key, Number(value / 10 / numCpu).toFixed(1), is_user, is_container, userid, this.state.podNameMapping) // usec/s → percent
         );
 
-        newState.topServicesMemory = n_biggest(this.cgroupMemoryNames, this.samples[10], 5).map(
-            ([key, value, is_user, is_container]) => cgroupRow(key, cockpit.format_bytes(value), is_user, is_container)
+        const topServicesMemory = n_biggest(this.cgroupMemoryNames, this.samples[10], 5);
+        newState.topServicesMemory = topServicesMemory.map(
+            ([key, value, is_user, is_container, userid]) => cgroupRow(key, cockpit.format_bytes(value), is_user, is_container, userid, this.state.podNameMapping)
         );
 
+        const notMappedContainers = topServicesMemory.concat(topServicesCPU).filter(([key, value, is_user, is_container, userid]) => is_container && !podNameInCache(this.state.podNameMapping, userid, key));
+        if (notMappedContainers.length !== 0) {
+            this.update_podman_name_mapping(notMappedContainers);
+        }
         this.setState(newState);
     }
 
+    /**
+     * Look up the container names using podman ps for the given cgroups. When
+     * the logged in user has administrative access we use sudo to call podman
+     * ps for that user.
+     */
+    update_podman_name_mapping = (cgroups) => {
+        const { privileged } = this.state;
+        const spawnPromise = (cmd, uid) => {
+            return new Promise((resolve, reject) => {
+                cockpit.spawn(cmd, { superuser: privileged ? "required" : null }).then(result => {
+                    resolve([result, uid]);
+                }).catch(err => reject(err));
+            });
```

Comment:
Unnecessary Promise wrapper, just return cockpit.spawn(). (or more precisely, the chained promise after .then()/.catch().